turbo/jsonrpc/otterscan_api.go

```diff
@@ -433,6 +458,111 @@ func (api *OtterscanAPIImpl) SearchTransactionsAfter(ctx context.Context, addr c
+	fromTxNum := -1
+	if fromBlockNum == 0 {
+		isLastPage = true
+	} else {
+		// Internal search code considers blockNum [including], so adjust the value
+		_txNum, err := rawdbv3.TxNums.Min(tx, fromBlockNum+1)
+		if err != nil {
+			return nil, err
+		}
+		fromTxNum = int(_txNum)
+	}
+
+	// ensure result union will always have at least pageSize+1 results in order to determine
+	// there is more results or this is the last page, at the same time avoiding unbounded
+	lookupSize := int(pageSize) + 1
+	itTo, err := tx.IndexRange(kv.TracesToIdx, addr[:], fromTxNum, -1, order.Asc, lookupSize)
+	if err != nil {
+		return nil, err
+	}
+	itFrom, err := tx.IndexRange(kv.TracesFromIdx, addr[:], fromTxNum, -1, order.Asc, lookupSize)
+	if err != nil {
+		return nil, err
+	}
+	txNums := iter.Union[uint64](itFrom, itTo, order.Asc, lookupSize)
+	txNumsIter := rawdbv3.TxNums2BlockNums(tx, txNums, order.Asc)
+
+	exec := txnExecutor(tx, chainConfig, api.engine(), api._blockReader, nil)
+	var blockHash common.Hash
+	var header *types.Header
+	txs := make([]*RPCTransaction, 0, pageSize)
+	receipts := make([]map[string]interface{}, 0, pageSize)
+	resultCount := uint16(0)
+
+	for txNumsIter.HasNext() {
+		txNum, blockNum, txIndex, isFinalTxn, blockNumChanged, err := txNumsIter.Next()
+		if err != nil {
+			return nil, err
+		}
+		if isFinalTxn {
+			continue
+		}
+
+		if blockNumChanged { // things which not changed within 1 block
+			if header, err = api._blockReader.HeaderByNumber(ctx, tx, blockNum); err != nil {
+				return nil, err
+			}
+			if header == nil {
+				log.Warn("[rpc] header is nil", "blockNum", blockNum)
+				continue
+			}
+			blockHash = header.Hash()
+			exec.changeBlock(header)
+		}
+
+		//fmt.Printf("txNum=%d, blockNum=%d, txIndex=%d, maxTxNumInBlock=%d,mixTxNumInBlock=%d\n", txNum, blockNum, txIndex, maxTxNumInBlock, minTxNumInBlock)
+		txn, err := api._txnReader.TxnByIdxInBlock(ctx, tx, blockNum, txIndex)
+		if err != nil {
+			return nil, err
+		}
+		if txn == nil {
+			continue
+		}
+		rawLogs, res, err := exec.execTx(txNum, txIndex, txn)
+		if err != nil {
+			return nil, err
+		}
+		rpcTx := NewRPCTransaction(txn, blockHash, blockNum, uint64(txIndex), header.BaseFee)
+		txs = append(txs, rpcTx)
+		receipt := &types.Receipt{
+			Type:              txn.Type(),
+			GasUsed:           res.UsedGas,
+			CumulativeGasUsed: res.UsedGas, // TODO: cumulative gas is wrong, wait for cumulative gas index fix
+			TransactionIndex:  uint(txIndex),
+			BlockNumber:       header.Number,
+			BlockHash:         blockHash,
+			Logs:              rawLogs,
+		}
+		if res.Failed() {
+			receipt.Status = types.ReceiptStatusFailed
+		} else {
+			receipt.Status = types.ReceiptStatusSuccessful
+		}
+
+		mReceipt := marshalReceipt(receipt, txn, chainConfig, header, txn.Hash(), true)
+		mReceipt["timestamp"] = header.Time
+		receipts = append(receipts, mReceipt)
+
+		resultCount++
+		if resultCount >= pageSize {
+			break
+		}
+	}
+	hasMore := txNumsIter.HasNext()
```

Comment:
Adding `lookupSize` to underlying iterators may break: `hasMore := txNumsIter.HasNext()`

Because: `if resultCount >= pageSize` will drain 100% of underlying iterators - we have 2 `continue` in this `for` loop - and `.HasNext()` will return `false` (only because both underlying iterators will touch `lookupSize` limit before `resultCount` bet to the point).

Reply:
Iterators are lazy - means "unbounded" is not so bad. 
FYI: having bound may be cheaper in case of remote "RPCDaemon without --datadir" - there records fetching by "own pages", in worst case it will fetch 1 full addition page (cheap enough, it will not increase amount of network calls).

Reply:
fixed, as the previous comment. it needs to be unbounded.